### ISSUE TITLE
get shapeInfo in ARM project using C++ interface

### ISSUE THREAD

**Issue by feng98** [dmlc/tvm#1454]
The ARM project need shape information of input data, so that have flexible parameter configuration. Is there any other convenient way to get the shape information of input data when use runtime in ARM? 
@tqchen @merrymercy

**Comment by srkreddy1238** [comment]
I feel we should have an graph runtime API just to query input and output dimensions.

@tqchen what do you think ?

**Comment by tqchen** [comment]
Sorry for delayed reply here, yes we want to add runtime interface to query the shape. The real question is what should it look like. 

- We could directly create an API that returns a tensor(ndarray) reference
    - Need to change the runtime
    - Do not work across the RPC boundary
- Create an information query function that returns dimension and each value of the shape
   - Need additional frontend wrapping
   - Works across RPC boundary

@dmlc/tvm-team what do you think

**Comment by srkreddy1238** [comment]
I vote for second option(frontend wrapping) to keep it common and available every where.

**Comment by kevinthesun** [comment]
Vote for 2 for RPC support.

**Comment by kazum** [comment]
+1 for the second one, for the same reason with @srkreddy1238 and @kevinthesun.

**Comment by feng98** [comment]
I vote for the second one.

**Comment by tmoreau89** [comment]
Same, option 2 - it's best to keep the RPC stable.

**Comment by zhreshold** [comment]
+1 for option 2.

**Comment by srkreddy1238** [comment]
What would be the final return for get_input_info ? 
How packed function args & return values ? receives: (name, field, index ) return : (integer) ?

Hope there would be similar for "get_output_info" too ..

**Comment by tqchen** [comment]
+1 for get_output_info. The get_input_info returns a PackedFunc, which takes in field, index and return the corresponding field.

**Comment by srkreddy1238** [comment]
@tqchen

**Comment by tqchen** [comment]
I think a bit more about it, maybe a better way is to enable RPC to return a remote NDArray in its return values, it needs a bit to refactor of RPC but doable, I will try a stab on this

**Comment by tqchen** [comment]
https://github.com/dmlc/tvm/pull/1610 support this feature, so we can directly do  get_input_array and get_output_array

**Comment by srkreddy1238** [comment]
@tqchen 
Do we leave existing ```get_output``` as is is and go with new API as ```get_input_array(index)``` and ```get_output_array(index)``` to return NDArray?

**Comment by tqchen** [comment]
yes, let us do that, however, let us make sure to redirect the calls if  can be implemented on top of another

**Comment by srkreddy1238** [comment]
#1659 for review.
